### PR TITLE
add distance conf filter

### PR DESCRIPTION
**Pull request by zjgemi**
## Summary by CodeRabbit

- **New Features**
	- Introduced a `filters` argument for configuration customization in existing functions.
	- Implemented new filtering classes for validating atomic configurations based on distance and geometric criteria, enhancing configuration selection options.

- **Tests**
	- Added unit tests for the new filtering classes to ensure robust functionality and validation of atomic configurations.

### CODE REVIEW

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+import shutil
+
+import dpdata
+import numpy as np
+
+from . import (
+    ConfFilter,
+)
+
+safe_dist_dict = {
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:11]: hard coding

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    "Pa": 2.8,
+    "U": 2.8,
+    "Np": 2.8,
+    "Pu": 2.8,
+    "Am": 2.8,
+    "Cm": 2.8,
+    "Cf": 2.3,
+}
+for k in safe_dist_dict:
+    safe_dist_dict[k] *= 0.441
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:111]: hard coding

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+            cell=cell,
+            pbc=(not nopbc),
+        )
+
+        cell = structure.get_cell()
+
+        if (
+            cell[1][0] > 1.732 * cell[1][1]
+            or cell[2][0] > 1.732 * cell[2][2]
+            or cell[2][1] > 1.732 * cell[2][2]
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:159]: hard coding of tolerance angles.

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        ):
+            print("Inclined box")
+            return False
+
+        a = cell[0][0]
+        b = cell[1][1]
+        c = cell[2][2]
+
+        if check_multiples(a, b, c, 5):
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:168]: hard coding of "5"

**Review by wanghan-iapcm — COMMENTED** [review]
To many hard coding. could you please provide interface allowing user specific parameters?

File: dpgen2/exploration/selector/conf_filter.py [hunk 5]
```diff
@@ -67,7 +67,7 @@ def check(
                     ff.check(
                         conf["coords"][ii],
                         conf["cells"][ii],
-                        conf["atom_types"],
+                        np.array([conf["atom_names"][t] for t in conf["atom_types"]]),
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/conf_filter.py:70]: we may want to provide both the atom types and atom names. 
or simply pass a frame of the dpdata.System. 
what do you think?

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+        self.custom_safe_dist = custom_safe_dist if custom_safe_dist is not None else {}
+        self.safe_dist_ratio = safe_dist_ratio
+        self.theta = theta
+        self.length_ratio = length_ratio
+
+    def check(
+        self,
+        coords: np.ndarray,
+        cell: np.ndarray,
+        atom_types: np.ndarray,
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:146]: it should be `atom_names`

Comment (wanghan-iapcm) [reply dpgen2/exploration/selector/distance_conf_filter.py:146]: an np.array of `str` looks strange..

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    def args() -> List[dargs.Argument]:
+        r"""The argument definition of the `ConfFilter`.
+
+        Returns
+        -------
+        arguments: List[dargs.Argument]
+            List of dargs.Argument defines the arguments of the `ConfFilter`.
+        """
+
+        doc_custom_safe_dist = "Custom safe distance for each element"
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:221]:
```suggestion
        doc_custom_safe_dist = "Custom safe distance (in unit of bohr) for each element"
```

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        doc_safe_dist_ratio = "The ratio multiplied to the safe distance"
+        doc_theta = "The threshold for the angle of the box"
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:223]:
```suggestion
        doc_theta = "The threshold for angles between the edges of the cell. If all angles are larger than this value the check is passed. "
```

File: dpgen2/exploration/selector/distance_conf_filter.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        doc_length_ratio = "The threshold for the length ratio of the box"
```
Comment (wanghan-iapcm) [inline dpgen2/exploration/selector/distance_conf_filter.py:224]:
```suggestion
        doc_length_ratio = "The threshold for the length ratio between the edges of the cell. If all length ratios are smaller than this value the check is passed"
```

File: tests/exploration/test_distance_conf_filter.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        distance_conf_filter = DistanceConfFilter()
+        valid = distance_conf_filter.check(
+            system["coords"][0],
+            system["cells"][0],
+            np.array([system["atom_names"][t] for t in system["atom_types"]]),
+            system.nopbc,
+        )
+        self.assertTrue(valid)
+
+    def test_invalid(self):
```
Comment (wanghan-iapcm) [inline tests/exploration/test_distance_conf_filter.py:75]: I do not think the test case covers all the filters.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: dpgen2/exploration/selector/conf_filter.py (hunk 5 above)
Comment (zjgemi) [reply dpgen2/exploration/selector/conf_filter.py:70]: Good idea

**Review by wanghan-iapcm — APPROVED** [review]

**Pull request merged**